flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFileDataIndexImplTest.java

```diff
@@ -0,0 +1,155 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid;
+
+import org.apache.flink.runtime.io.network.partition.hybrid.HsFileDataIndex.ReadableRegion;
+import org.apache.flink.runtime.io.network.partition.hybrid.HsFileDataIndex.SpilledBuffer;
+import org.apache.flink.util.TestLoggerExtension;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for {@link HsFileDataIndexImpl}. */
+@ExtendWith(TestLoggerExtension.class)
+class HsFileDataIndexImplTest {
+    private static final int NUM_SUBPARTITIONS = 3;
+
+    private HsFileDataIndexImpl hsDataIndex;
+
+    @BeforeEach
+    void before() {
+        hsDataIndex = new HsFileDataIndexImpl(NUM_SUBPARTITIONS);
+    }
+
+    /**
+     * If a subpartition has no spilled buffer added to the index, all the query results are {@link
+     * Optional#empty()}.
+     */
+    @Test
+    void testGetReadableRegionForEmptySubpartition() {
+        // Only add buffers for subpartition 0.
+        hsDataIndex.addBuffers(createSpilledBuffers(0, Arrays.asList(0, 10)));
+        hsDataIndex.markBufferReadable(0, 0);
+        hsDataIndex.markBufferReadable(0, 10);
+        // Different subpartitions are independent and will not affect each other.
+        assertThat(hsDataIndex.getReadableRegion(1, 0)).isNotPresent();
+        assertThat(hsDataIndex.getReadableRegion(2, 10)).isNotPresent();
+    }
+
+    /**
+     * If the buffer index with the corresponding subpartition is not exists in the data index, the
+     * return value should be {@link Optional#empty()}.
+     */
+    @Test
+    void testGetReadableRegionWhenBufferIsNotExists() {
+        final int targetSubpartition = 0;
+        // spilled buffers for subpartition 0: (0-4, 0-5), (0-7, 0-8)
+        // note that: 0-4 indicate buffer within subpartition 0 and buffer index is 4.
+        List<SpilledBuffer> spilledBuffers =
+                createSpilledBuffers(targetSubpartition, Arrays.asList(4, 5, 7, 8));
+
+        hsDataIndex.addBuffers(spilledBuffers);
+
+        assertThat(hsDataIndex.getReadableRegion(targetSubpartition, 6)).isNotPresent();
+    }
```

Comment:
Seems `testGetReadableRegionForEmptySubpartition` is a special case of `testGetReadableRegionWhenBufferIsNotExists`. They probably can be combined into one case.

Reply:
merged them together.